tools/wake-format/emitter.cpp

```diff
@@ -51,69 +44,48 @@ static bool is_expression(uint8_t type) {
 }
 
 void Emitter::walk_rhs(wcl::doc_builder& bdr, ctx_t ctx, CSTElement& node) {
-  wcl::doc doc = space().concat(walk_node(ctx.sub(bdr), node));
-  if (fits(bdr, ctx, doc) && !requires_nl(ctx, node)) {
-    bdr.append(doc);
-  } else {
-    ctx_t n_ctx = ctx.nest();
-    bdr.append(newline(n_ctx));
-    bdr.append(walk_node(n_ctx.sub(bdr), node));
+  auto nested_fmt = formatter().nest(formatter().walk(WALK(walk_node)));
+  if (requires_nl(ctx, node)) {
+    bdr.append(nested_fmt.format(ctx, node, false));
+    node.nextSiblingElement();
+    return;
   }
-  node.nextSiblingElement();
-}
-
-wcl::doc Emitter::newline(ctx_t ctx) {
-  wcl::doc_builder bdr;
 
-  bdr.append("\n");
-  for (size_t i = 0; i < ctx.nest_level; i++) {
-    bdr.append(space(space_per_indent));
+  wcl::doc doc = formatter().space().walk(WALK(walk_node)).format(ctx, node, false);
+  if (fits(bdr, ctx, doc)) {
+    bdr.append(doc);
+    node.nextSiblingElement();
+    return;
   }
 
-  return std::move(bdr).build();
-}
-
-wcl::doc Emitter::space(uint8_t count) {
-  wcl::doc_builder bdr;
-  for (uint8_t i = 0; i < count; i++) {
-    bdr.append(" ");
-  }
-  return std::move(bdr).build();
+  bdr.append(nested_fmt.format(ctx, node, false));
+  node.nextSiblingElement();
 }
 
 wcl::doc Emitter::layout(CST cst) {
   ctx_t ctx;
-  return walk(ctx, cst.root()).concat(newline(ctx));
+  return formatter().walk(WALK(walk)).format(ctx, cst.root());
 }
 
 wcl::doc Emitter::walk(ctx_t ctx, CSTElement node) {
   wcl::doc_builder bdr;
 
-  for (CSTElement child = node.firstChildElement(); !child.empty(); child.nextSiblingElement()) {
-    // remove optional whitespace
-    if (child.id() == TOKEN_WS) {
-      continue;
-    }
-
-    if (child.id() == TOKEN_NL) {
-      bdr.append(newline(ctx));
-      continue;
-    }
-
-    if (child.id() == TOKEN_COMMENT) {
-      bdr.append(walk_token(ctx, child));
-      continue;
-    }
-
-    assert(child.isNode());
-
-    bdr.append(walk_node(ctx, child));
-    bdr.append(newline(ctx));
-  }
-
-  bdr.undo();
-
-  return std::move(bdr).build();
+  // clang-format off
+  auto body_fmt = formatter()
```

Comment:
I'm confused about what this does? It looks like this

```
if (!is_whitespace) {
  continue
}

if (is_newline) {
  emit_newline
  continue
}

if (is_comment) {
  emit_comment
  cotinue
}

otherwise walk the node
```

ok now that I wrote that out I get it....hmm this should be better somehow. What about this

```
auto extra_fmt =
  formatter()
  .fmt_if(TOKEN_WS, formatter())
  .fmt_if(TOKEN_NL, formatter().newline(space_per_indent))
  .fmt_if(TOKEN_COMMENT, formatter().walk(WALK(walk_token)));

auto body_fmy =
  formatter()
  .fmt_if_else(
    {TOKEN_WS, TOKEN_NL, TOKEN_COMMENT},
    extra_fmt,
    formatter().walk(WALK(walk_node)).newline(space_per_indent))));
```

Reply:
yeah, that is a bit clearer, will need a new lever to do the `{TOKEN_WS, TOKEN_NL, TOKEN_COMMENT}` stuff though

Reply:
It's just the bitset thing again